### PR TITLE
Autocompletion for plugins

### PR DESCRIPTION
**Pull request by wargio**
Example of plugin with autocompletion: https://github.com/wargio/r2dec-js/pull/65

### CODE REVIEW

**Comment by radare** [comment]
Can you add _del() and call those 3 functions from an r2 command? (we need to decide which one)

* I propose `!!!`

**Comment by wargio** [comment]
i'm ok with `!!!`

**Comment by radare** [comment]
yep, that sounds good. can you update the PR with that command?

**Comment by wargio** [comment]
Do not merge yet, PLS!!

File: libr/core/cmd.c [hunk 1]
```diff
@@ -1378,20 +1378,111 @@ static int cmd_env(void *data, const char *input) {
 	return ret;
 }
 
+static inline const char* next_arg(const char *input) {
+	while (input && *input && IS_WHITESPACE (*input)) {
```
Comment (radare) [inline libr/core/cmd.c]: this function is a copy of r_str_trim_ro()

File: libr/core/cmd.c [hunk 2]
```diff
@@ -1378,20 +1378,111 @@ static int cmd_env(void *data, const char *input) {
...
+		input++;
+	}
+	return input;
+}
+
+static inline const char* next_whitespace(const char *input) {
```
Comment (radare) [inline libr/core/cmd.c]: uhm maybe move that logic into util/str

File: libr/core/cmd.c [hunk 3]
```diff
@@ -1378,20 +1378,111 @@ static int cmd_env(void *data, const char *input) {
...
+			eprintf ("Added '%s' into the autocomplete dict.\n", arg);
+			return;
+		} else if ((!input || !*input) && a && remove) {
+			eprintf("Remove from %s\n", b->cmd);
+			if (r_core_autocomplete_remove (b, arg)) {
+				eprintf ("Removed '%s' into the autocomplete dict.\n", arg);
+			} else {
+				eprintf ("Removing '%s' from the dict is forbidden.\n", arg);
+			}
+			return;
```
Comment (radare) [inline libr/core/cmd.c]: dont else after return

File: libr/core/core.c [hunk 4]
```diff
@@ -914,8 +914,78 @@ static int autocomplete_pfele (RCore *core, char *key, char *pfx, int idx, char
...
+		if (!strncmp (radare_argv[i], line->buffer.data, line->buffer.index)) {
+			tmp_argv[j++] = radare_argv[i];
+		}
+	}
+	tmp_argv[j] = NULL;
+	line->completion.argc = j;
+	line->completion.argv = tmp_argv;
+}
+
+static inline const char* next_whitespace(const char *input) {
```
Comment (radare) [inline libr/core/core.c]: dupe

File: libr/core/core.c [hunk 5]
```diff
@@ -914,8 +914,78 @@ static int autocomplete_pfele (RCore *core, char *key, char *pfx, int idx, char
...
+	while (input && *input && !IS_WHITESPACE (*input)) {
+		input++;
+	}
+	return input;
+}
+
+static inline const char* next_arg(const char *input) {
```
Comment (radare) [inline libr/core/core.c]: more dupe

File: libr/core/core.c [hunk 6]
```diff
@@ -914,8 +914,78 @@ static int autocomplete_pfele (RCore *core, char *key, char *pfx, int idx, char
...
+		if (!e || (e - p) >= 256 || e == p) {
+			return NULL;
+		}
+		memcpy (arg, p, e - p);
+		arg[e - p] = 0;
+		tmp = r_core_autocomplete_find (plugin, arg);
+		if (tmp && *(p + tmp->length) == ' ') {
+			p = next_arg (p + tmp->length);
+			plugin = tmp;
+		} else if(tmp) {
```
Comment (radare) [inline libr/core/core.c]: missing space and dont else after break

File: libr/core/cmd.c (hunk 3 above)
Comment (wargio) [reply libr/core/cmd.c]: no. i need that return there.

**Comment by radare** [comment]
<3

**Comment by wargio** [comment]
https://asciinema.org/a/HDyXRyNvImU39ZvOgChSuNGn3

**Comment by radare** [comment]
<3

**Comment by MaskRay** [comment]
Old `e cfg.newtab=true` prints help (`a TAB` executes and prints `a?`) when you hit a tab. Now the logic gets lost. Do you know how to reinstate the behavior? r2 has so many commands.

https://github.com/radare/radare2/blob/master/libr/core/core.c#L1561

		} else if (!find_autocomplete (line)) { // find_autocomplete returns true
			int i, cfg_newtab = r_config_get_i (core->config, "cfg.newtab");

**Pull request merged**